src/strings/ops.c

```diff
@@ -1220,9 +1220,18 @@ MVMString * MVM_string_join(MVMThreadContext *tc, MVMString *separator, MVMObjec
 
             /* Add separator if needed. */
             if (i > 0) {
+                /* If there's no separator and one piece is The Empty String we
+                 * have to be extra careful about concat stability */
+                if (sgraphs == 0 && MVM_string_graphs(tc, piece) == 0) {
```

Comment:
You can use MVM_string_graphs_nocheck here since MVM_string_graphs is already run on each piece in the loop above.